src/Microsoft.VisualStudio.Threading.Analyzers.Tests/VSTHRD112AvoidNullReturnInNonAsyncTaskMethodAnalyzerTests.cs

```diff
@@ -0,0 +1,137 @@
+﻿namespace Microsoft.VisualStudio.Threading.Analyzers.Tests
+{
+    using System.Threading.Tasks;
+    using Xunit;
+    using Verify = CSharpCodeFixVerifier<VSTHRD112AvoidNullReturnInNonAsyncTaskMethodAnalyzer, CodeAnalysis.Testing.EmptyCodeFixProvider>;
+
+    public class VSTHRD112AvoidNullReturnInNonAsyncTaskMethodAnalyzerTests
+    {
+        [Fact]
+        public async Task TaskOfTReturnsNull_Diagnostic()
+        {
+            var test = @"
+using System.Threading.Tasks;
+
+class Test
+{
+    public Task<object> GetTaskObj()
+    {
+        return null;
+    }
+}
+";
+            await new Verify.Test
+            {
+                TestCode = test,
+                ExpectedDiagnostics = { Verify.Diagnostic().WithSpan(8, 9, 8, 21), },
+            }.RunAsync();
+        }
+
+        [Fact]
+        public async Task TaskReturnsNull_Diagnostic()
+        {
+            var test = @"
+using System.Threading.Tasks;
+
+class Test
+{
+    public Task GetTask()
+    {
+        return null;
+    }
+}
+";
+            await new Verify.Test
+            {
+                TestCode = test,
+                ExpectedDiagnostics = { Verify.Diagnostic().WithSpan(8, 9, 8, 21), },
+            }.RunAsync();
+        }
+
+        [Fact]
+        public async Task AsyncReturnsNull_NoDiagnostic()
+        {
+            var test = @"
+using System.Threading.Tasks;
+
+class Test
+{
+    public async Task<object> GetTaskObj()
+    {
+        return null;
+    }
+}
+";
+            await new Verify.Test
+            {
+                TestCode = test,
+            }.RunAsync();
+        }
+
+        [Fact]
+        public async Task VariableIsNullAndReturned_NoDiagnostic_FalseNegative()
+        {
+            var test = @"
+using System.Threading.Tasks;
+
+class Test
+{
+    public Task<object> GetTaskObj()
+    {
+        Task<object> o = null;
+        return o;
```

Comment:
We would need dataflow analysis to detect that there is a path returning a null object.

Reply:
Understood. I won't require that we fix this. :)